autorest/codegen/templates/operations_container_mixin.py.jinja2

```diff
@@ -3,10 +3,80 @@
 {{ code_model.options['license_header'] }}
 {{ imports }}
 
-{{ "    " if not async_mode }}T = TypeVar('T')
-{{ "    " if not async_mode }}ClsType = Optional[Callable[[PipelineResponse[HttpRequest, {{ "Async" if async_mode else "" }}HttpResponse], T, Dict[str, Any]], Any]]
+{% if not async_mode %}
+    {# since we define ClsType outside of the typing section in async mode, we will define it below the AbstractServiceClient in async mode #}
+    T = TypeVar('T')
+    ClsType = Optional[Callable[[PipelineResponse[HttpRequest, {{ "Async" if async_mode else "" }}HttpResponse], T, Dict[str, Any]], Any]]
 
-class {{ operation_group.class_name }}{{ object_base_class }}:
+{% endif %}
+    class AbstractServiceClient(object):
```

Comment:
Shall this be a protocol?
https://www.python.org/dev/peps/pep-0544/#defining-a-protocol

Reply:
Shall we have this protocol in an external place "autorest-like client protocols"? azure-core? elsewhere? @johanste